PermissionVault/contracts/PermissionVault.sol

```diff
@@ -0,0 +1,348 @@
+    _setupRole(CONTROLLER_ROLE, controller);
+
+    emit AddController(controller);
+  }
+
+  /**
+   * @notice The owner can use this function to revoke the CONTROLLER_ROLE from
+   * an address. The controller parameter is the address that will have the
+   * CONTROLLER_ROLE revoked.
+   * @dev Callable by owner
+   * @param controller controller address
+   */
+  function removeController(address controller) external onlyOwner {
+    revokeRole(CONTROLLER_ROLE, controller);
+
+    emit RemoveController(controller);
+  }
+
+  /**
+   * @notice A controller can use this function to deposit Ether into the
+   * contract.
+   */
+  function depositEther() external payable {
+    if (msg.value == 0) {
+      revert ZeroAmount();
+    }
+
+    emit DepositEther(msg.sender, msg.value);
+  }
+
+  /**
+   * @notice A controller can use this function to withdraw Ether from the contract.
+   * @dev Callable by controller
+   * @param to the address that will receive the withdrawn Ether
+   * @param amount the amount of Ether to withdraw
+   */
+  function withdrawEther(
+    address to,
+    uint256 amount
+  ) external nonReentrant onlyController {
+    if (amount > address(this).balance) {
+      revert NotEnoughBalance();
+    }
+
+    (bool success, ) = to.call{value: amount}("");
+    require(success, "Transfer ether failed");
+
+    emit WithdrawEther(msg.sender, to, amount);
+  }
+
+  /**
+   * @notice A controller can use this function to deposit ERC20 tokens into the contract.
+   * @param token the ERC20 token contract address
+   * @param amount the amount of tokens to deposit
+   */
+  function depositERC20(IERC20 token, uint256 amount) external {
+    if (amount == 0) {
+      revert ZeroAmount();
+    }
+    if (token.balanceOf(msg.sender) < amount) {
+      revert NotEnoughBalance();
+    }
+
+    token.safeTransferFrom(msg.sender, address(this), amount);
+    emit DepositERC20(msg.sender, token, amount);
+  }
+
+  /**
+   * @notice A controller can use this function to withdraw ERC20 tokens from the contract.
+   * @param to the address that will receive the withdrawn tokens
+   * @param token the ERC20 token contract address
+   * @param amount the amount of tokens to withdraw
+   */
+  function withdrawERC20(
+    address to,
+    IERC20 token,
+    uint256 amount
+  ) external onlyController {
+    if (amount > token.balanceOf(address(this))) {
+      revert NotEnoughBalance();
+    }
+
+    token.safeTransfer(to, amount);
+    emit WithdrawERC20(msg.sender, to, token, amount);
+  }
+
+  /**
+   * @notice A controller can use this function to deposit ERC721 tokens into the contract.
+   * @param token the ERC721 token contract address
+   * @param id the token ID to deposit
+   */
+  function depositERC721(IERC721 token, uint256 id) external {
+    if (token.ownerOf(id) != msg.sender) {
+      revert NotOwnerOf();
+    }
+    token.safeTransferFrom(msg.sender, address(this), id);
```

Comment:
Please check approve allowance before safeTransferFrom

Reply:
don't need it